### ISSUE TITLE
Questions about the training details

### ISSUE THREAD

**Issue by Code-kunkun** [facebookresearch/VidDetours#2]
Thanks for your excellent work!!! I have some questions about the training procedure.
(1) Does the retrieval procedure and the localization procedure are separate? In other words, do we need to first train a model to finish the retrieval stage, and then train another model to complete the localization process?
(2) According to the given scripts, it seems that the retrieval model is based on llama2-7b, while the localization model is based on llama2-13b. Why did you choose two different versions? And could you tell me how much training time is needed for each of these two stages?
(3) When will the checkpoint be released?

**Comment by thechargedneutron** [comment]
Hi, thanks for the questions

1. Localization and retrieval are two separate experiments with two separate architecture and inputs.

2. We do both experiments with 13B. It is faster to train with 7B so we had those runs. Overall, we were getting big gains with both 7B and 13B in retrieval. For localization, 13B gave more gain, so we report both numbers with 13B. If you finetune the whole network, it takes around 1.5 days. If you parallelize this on multiple nodes, it will be faster. In case you do not have the required compute, you can just train the mapper and that also gives good results.

3. I do not have Meta's permission yet to release the checkpoint. I need to re-train the model after cleaning the data, so it will take some time depending on my schedule. Unfortunately, I cannot commit a timeline.

Please let me know if you have more questions.

**Comment by Code-kunkun** [comment]
When I run a retrieval experiment based on Llama2-7b, I find that even with the batch size to 1, I still get an out-of-memory error. I am using an A100 80GB.

**Comment by thechargedneutron** [comment]
I ran the experiments with 8 A100 80GB GPUs. Maybe the model does not fit on a single GPU. Are you able to run the vanilla Llama-2-7b on one GPU? Our retrieval model adds some parameters on top of LLama-2-7b so if the vanilla Llama-2-7b runs on one GPU then you can try reducing the number of parameters? 

Also, which stage of deepspeed are you using?

**Comment by Code-kunkun** [comment]
I use the following script. When I freeze the backbone, it can run properly, while when I open Backbone training, it will lead to out-of-memory.
<img width="984" alt="image" src="https://github.com/facebookresearch/VidDetours/assets/58265850/b589e758-ac30-421d-b63e-042b512e7bc3">

**Comment by thechargedneutron** [comment]
Can you try using one of the deepspeed configs used in LLaVA codebase: https://github.com/haotian-liu/LLaVA/blob/main/scripts/pretrain.sh#L16

This uses zero2, maybe you should be able to fit if you use zero3. The training will naturally be slower but you will be able to fit. Can you try this and let me know if it works or not?

**Comment by Code-kunkun** [comment]
When I set the `retrieval_aggregator` to `average`, it seems to work well.
<img width="969" alt="image" src="https://github.com/facebookresearch/VidDetours/assets/58265850/e8eeea2b-1dc4-41b3-b25f-fcf13c74cbf4">
But when I set the `retrieval_aggregator` to `transformer`, I encounter the following error. The error is reported during the self-attention processing of the transformer encoder layer in the retrieval head. I have checked the dimensions of the input and there seems to be no problem.😭
<img width="1192" alt="image" src="https://github.com/facebookresearch/VidDetours/assets/58265850/93ef285a-6ced-416b-873e-30dd154c1d21">
when I set the `retrieval_aggregator` to `transformer`, and use `zero2`, the error disappears, but I can only fine-tune the model with Lora, otherwise it will lead to out-of-memory. 

The code doesn't seem to explain how Recall@K is calculated in the retrieval stage. Is it sorted in descending order based on the predicted probabilities and then the ranking of the target is calculated?

**Comment by thechargedneutron** [comment]
> When I set the retrieval_aggregator to average, it seems to work well.

Yes, this is because average does not add any additional parameter and the overflow is just avoided.

> I encounter the following error. The error is reported during the self-attention processing of the transformer encoder layer in the retrieval head. I have checked the dimensions of the input and there seems to be no problem.

Yes, this looks like a deepspeed error - most likely changing the deepspeed version should solve this.

> when I set the retrieval_aggregator to transformer, and use zero2, the error disappears, but I can only fine-tune the model with Lora, otherwise it will lead to out-of-memory.

You can try the full Lora training, we were getting very good results with only visual mapper training. We only did the full fine-tuning at the end to further boost the performance.

> The code doesn't seem to explain how Recall@K is calculated in the retrieval stage. Is it sorted in descending order based on the predicted probabilities and then the ranking of the target is calculated?

Yes, this is correct.

Unfortunately, I ran this code on eight A100 GPUs and hence, running it on one GPU might be tricky. If you want to do full fine-tuning, you may want to reduce the retrieval head that I have added on top of LLAMA-2-7b.

**Comment by Code-kunkun** [comment]
Have you ever tried training with fp16? It seems there are some bugs when I train with fp16.

And what is the approximate lowest value your localization loss can decrease to? In my experiments, I've found that the localization loss doesn't decrease significantly.

**Comment by thechargedneutron** [comment]
@Code-kunkun can you show what error are you getting for fp16? These configurations are also hardware dependent. One way to go about it is to try running LLaVA and use the same parameters for VidDetours as well. We do have additional parameters and that might result in additional OOM issues but then using deepspeed offloading should work.

In the meantime, I will try to run this code on one A100 GPU to see if I see OOM.

**Comment by Code-kunkun** [comment]
![image](https://github.com/facebookresearch/VidDetours/assets/58265850/6b352af1-6abe-4833-8aba-20b5ceb44e47)
In the test set, there are some queries where the video2 start second is greater than the end second. Is this considered an annotation error? Should these problematic samples be ignored?

**Comment by thechargedneutron** [comment]
This is weird. Thanks for pointing this out. Can you check if [end, start] makes sense in such cases? But given that number of such annotations is low, it can be ignored.

**Comment by AAnminy** [comment]
同学你好，很抱歉在你的issue下回复无关的东西。我也在follow这项工作，请问你在跑代码的时候遇到了这个问题吗？
![屏幕截图 2024-12-02 204006](https://github.com/user-attachments/assets/8f32c470-3158-4cc0-8455-6e1b909edd42)

**Comment by Code-kunkun** [comment]
你可以试试安装4.42.3版本的transformers，我之前好像装的是这个版本。

**Comment by AAnminy** [comment]
> 你可以试试安装4.42.3版本的transformers，我之前好像装的是这个版本。

我试了试安装4.42.3还是有同样的问题出现，我直接自己写了那个函数，把它加在modeling_bloom里了😹，真的很奇怪，这个代码我就是跑不动。能问问你跑成功了吗？你train出来效果是否和paper里的描述差别多呢？🤓

**Comment by Code-kunkun** [comment]
直接运行的话，好像是会遇到奇奇怪怪的报错，我当时是自己重写了下，然后改了下网络设计，效果会比论文里的稍微高一些。

**Comment by AAnminy** [comment]
> 直接运行的话，好像是会遇到奇奇怪怪的报错，我当时是自己重写了下，然后改了下网络设计，效果会比论文里的稍微高一些。

如果方便的话，我们可以加一下vx吗？我的微信号是：WLUCKYGODDESSA（我前天给你的学校邮箱发了一封邮件，里面有一些我的个人信息🥹）
